### PR TITLE
fix(spec): bound vector dims and identifiers at verification (ARN-217)

### PR DESCRIPTION
**Pull request by nerdsane**
## ARN-217 — Vector declaration resource bounds

Fixes the unbounded vector dimensionality in ARN-217.

### The vulnerability

`VectorDecl.dims` had **no upper bound** (`crates/temper-spec/.../types.rs` —
`pub dims: usize`; `validate_vector_decls` only checked `dims > 0`). A spec could
declare `[[vector]] … dims = 100_000_000`, which the verification cascade
accepted. Query work and allocation are `rows × dims`: the row count was already
bounded (ADR-0155's `candidate_budget`) and `unpack_f32_le` already rejects
non-finite values, but with the dimensionality unbounded a single spec could
still force multi-gigabyte per-query allocation and dot-product work (400 MB per
100 M-dim vector blob), and the candidate scan materializes each blob before the
row-count check.

### The fix (ADR-0159, verification-only)

`validate_vector_decls` now rejects, at spec verification (fail-closed):
- `dims > MAX_VECTOR_DIMS` (16 384 — well above any real embedding model, which
  top out in the low thousands, so no legitimate spec is rejected);
- `name` / `property` / `model_property` identifiers longer than
  `MAX_VECTOR_IDENT_LEN` (256).

`dims` is a spec property enforced **once**, so every storage backend inherits
the bound with no per-backend code. Per-query work is then bounded by
construction — `work ≤ candidate_budget × MAX_VECTOR_DIMS`, two explicit
constants — and each pre-check blob is now at most `MAX_VECTOR_DIMS × 4` bytes,
closing the "materialize before the count check" concern without touching the
query path.

A separate per-query `candidate_count × dims` element budget was considered and
rejected (see ADR): it is redundant given the two caps, and sizing its constant
without wrongly rejecting a legitimate kNN query depends on the deployment's
`candidate_budget`.

### Tests

`crates/temper-spec/src/automaton/parser_test.rs` (`vector_validation`):
`rejects_oversized_dims` (100 M dims → rejected), `accepts_dims_at_the_maximum`
(the cap itself verifies — no legit spec rejected), `rejects_oversized_identifier`.
Committed first as a RED test (100 M-dim spec verified successfully).

### Live local E2E

See the `ARENA SHIPPABLE` comment for before/after.

<details open><summary><h3>Greptile Summary</h3></summary>

This PR closes ARN-217 by adding two upper bounds in `validate_vector_decls`: `dims > MAX_VECTOR_DIMS` (16 384) and identifier length `> MAX_VECTOR_IDENT_LEN` (256) for the `name`, `property`, and `model_property` fields of every `[[vector]]` declaration. Combined with ADR-0155's `candidate_budget`, query work is now bounded by construction (`rows × dims` with both sides capped).

- **`parser.rs`**: two new module-level constants with doc comments, two new rejection branches inserted between the `dims == 0` guard and the `METRICS` check, following the existing `validate_vector_decls` ordering.
- **`parser_test.rs`**: three tests added to `vector_validation` — rejection at 100 M dims, acceptance at the boundary value (referencing the constant directly), and rejection when `name` exceeds 256 chars.
- **`docs/adrs/0159-vector-resource-bounds.md`**: ADR documenting the decision, consequences, DST compliance rationale, and the rejected per-query element-budget alternative.
</details>

<details open><summary><h3>Confidence Score: 4/5</h3></summary>

The change touches only `validate_vector_decls` in `temper-spec` — no runtime, query, or simulation-visible path is modified. Both new checks are pure integer and string-length comparisons that short-circuit with an error; the happy path is unchanged.

The core fix is sound: the two constants are well-chosen, the checks are in the right place, and the ADR thoroughly documents the tradeoffs. Two things keep this short of fully clean: the `metric` field is absent from the identifier-length loop, and the `property`/`model_property` branches of the length check have no direct test.

`parser.rs` (lines 711–723 — the identifier-length loop missing `metric`) and `parser_test.rs` (no test for over-long `property`/`model_property` via a state variable with a long name).
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| crates/temper-spec/src/automaton/parser.rs | Adds `MAX_VECTOR_DIMS` (16 384) and `MAX_VECTOR_IDENT_LEN` (256) constants plus enforcement in `validate_vector_decls`; the `metric` field is not covered by the new identifier-length loop, and the `name` length check fires after `name` has already been embedded in earlier error messages. |
| crates/temper-spec/src/automaton/parser_test.rs | Three new tests covering over-size dims, boundary acceptance, and over-long name; `property`/`model_property` identifier bounds are not directly exercised by any test. |
| docs/adrs/0159-vector-resource-bounds.md | Well-structured ADR documenting the rationale, decision, consequences, DST compliance note, and rejected alternatives for the vector resource bounds fix. |

</details>

<details open><summary><h3>Flowchart</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[Spec submission] --> B[TOML parse]
    B --> C[validate_vector_decls loop]
    C --> D{Duplicate name?}
    D -- yes --> ERR1[Reject: declared twice]
    D -- no --> E{property in state vars?}
    E -- no --> ERR2[Reject: undeclared property]
    E -- yes --> F{model_property in state vars?}
    F -- no --> ERR3[Reject: undeclared model_property]
    F -- yes --> G{dims == 0?}
    G -- yes --> ERR4[Reject: dims must be > 0]
    G -- no --> H{dims > MAX_VECTOR_DIMS 16 384?}
    H -- yes --> ERR5[Reject: dims exceeds maximum NEW]
    H -- no --> I{any identifier len > MAX_VECTOR_IDENT_LEN 256?}
    I -- yes --> ERR6[Reject: identifier exceeds maximum NEW]
    I -- no --> J{metric in cosine dot l2?}
    J -- no --> ERR7[Reject: unknown metric]
    J -- yes --> K{More vector decls?}
    K -- yes --> C
    K -- no --> OK[Spec accepted]
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A[Spec submission] --> B[TOML parse]
    B --> C[validate_vector_decls loop]
    C --> D{Duplicate name?}
    D -- yes --> ERR1[Reject: declared twice]
    D -- no --> E{property in state vars?}
    E -- no --> ERR2[Reject: undeclared property]
    E -- yes --> F{model_property in state vars?}
    F -- no --> ERR3[Reject: undeclared model_property]
    F -- yes --> G{dims == 0?}
    G -- yes --> ERR4[Reject: dims must be > 0]
    G -- no --> H{dims > MAX_VECTOR_DIMS 16 384?}
    H -- yes --> ERR5[Reject: dims exceeds maximum NEW]
    H -- no --> I{any identifier len > MAX_VECTOR_IDENT_LEN 256?}
    I -- yes --> ERR6[Reject: identifier exceeds maximum NEW]
    I -- no --> J{metric in cosine dot l2?}
    J -- no --> ERR7[Reject: unknown metric]
    J -- yes --> K{More vector decls?}
    K -- yes --> C
    K -- no --> OK[Spec accepted]
```

</a>
</details>

<a href="https://app.greptile.com/ide/claude-code?prompt=Fix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Acrates%2Ftemper-spec%2Fsrc%2Fautomaton%2Fparser.rs%3A711-715%0AThe%20%60metric%60%20field%20is%20absent%20from%20the%20identifier-length%20loop%2C%20so%20a%20pathological%20metric%20string%20%28e.g.%2C%20a%201%20MB%20value%29%20passes%20through%20the%20new%20bounds%20check%20unimpeded%20and%20ends%20up%20verbatim%20in%20the%20error%20message%20at%20the%20subsequent%20%60METRICS.contains%60%20check.%20While%20%60metric%60%20is%20constrained%20to%20exactly%20three%20valid%20values%20at%20that%20step%2C%20adding%20it%20here%20keeps%20the%20bounds%20logic%20self-contained%20and%20consistent%20with%20the%20stated%20goal%20of%20bounding%20all%20%60%5B%5Bvector%5D%5D%60%20identifiers.%0A%0A%60%60%60suggestion%0A%20%20%20%20%20%20%20%20for%20%28label%2C%20ident%29%20in%20%5B%0A%20%20%20%20%20%20%20%20%20%20%20%20%28%22name%22%2C%20%26vec_decl.name%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%28%22property%22%2C%20%26vec_decl.property%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%28%22model_property%22%2C%20%26vec_decl.model_property%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%28%22metric%22%2C%20%26vec_decl.metric%29%2C%0A%20%20%20%20%20%20%20%20%5D%20%7B%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%202%0Acrates%2Ftemper-spec%2Fsrc%2Fautomaton%2Fparser_test.rs%3A246-261%0A**Identifier%20bounds%20for%20%60property%60%2F%60model_property%60%20are%20untested.**%20The%20comment%20correctly%20explains%20that%20an%20over-long%20string%20that%20does%20not%20match%20any%20declared%20state%20variable%20is%20caught%20by%20the%20state-var%20existence%20check%20first.%20However%2C%20the%20check%20on%20lines%20716%E2%80%93722%20is%20also%20exercised%20when%20a%20state%20variable%20itself%20has%20a%20name%20longer%20than%20256%20chars%20and%20is%20then%20referenced%20by%20%60property%60%20or%20%60model_property%60.%20Adding%20a%20companion%20test%20that%20declares%20such%20a%20state%20variable%20and%20then%20references%20it%20covers%20this%20branch%20directly%2C%20preventing%20a%20future%20refactor%20from%20silently%20moving%20the%20existence%20check%20before%20the%20length%20check%20without%20a%20failing%20test%20to%20signal%20the%20regression.%0A%0A&repo=nerdsane%2Ftemper&pr=358&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=6">Fix All in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22nerdsane%2Ftemper%22%20on%20the%20existing%20branch%20%22claude%2Farn-217-vector-bounds%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22claude%2Farn-217-vector-bounds%22.%0A%0AFix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Acrates%2Ftemper-spec%2Fsrc%2Fautomaton%2Fparser.rs%3A711-715%0AThe%20%60metric%60%20field%20is%20absent%20from%20the%20identifier-length%20loop%2C%20so%20a%20pathological%20metric%20string%20%28e.g.%2C%20a%201%20MB%20value%29%20passes%20through%20the%20new%20bounds%20check%20unimpeded%20and%20ends%20up%20verbatim%20in%20the%20error%20message%20at%20the%20subsequent%20%60METRICS.contains%60%20check.%20While%20%60metric%60%20is%20constrained%20to%20exactly%20three%20valid%20values%20at%20that%20step%2C%20adding%20it%20here%20keeps%20the%20bounds%20logic%20self-contained%20and%20consistent%20with%20the%20stated%20goal%20of%20bounding%20all%20%60%5B%5Bvector%5D%5D%60%20identifiers.%0A%0A%60%60%60suggestion%0A%20%20%20%20%20%20%20%20for%20%28label%2C%20ident%29%20in%20%5B%0A%20%20%20%20%20%20%20%20%20%20%20%20%28%22name%22%2C%20%26vec_decl.name%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%28%22property%22%2C%20%26vec_decl.property%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%28%22model_property%22%2C%20%26vec_decl.model_property%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%28%22metric%22%2C%20%26vec_decl.metric%29%2C%0A%20%20%20%20%20%20%20%20%5D%20%7B%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%202%0Acrates%2Ftemper-spec%2Fsrc%2Fautomaton%2Fparser_test.rs%3A246-261%0A**Identifier%20bounds%20for%20%60property%60%2F%60model_property%60%20are%20untested.**%20The%20comment%20correctly%20explains%20that%20an%20over-long%20string%20that%20does%20not%20match%20any%20declared%20state%20variable%20is%20caught%20by%20the%20state-var%20existence%20check%20first.%20However%2C%20the%20check%20on%20lines%20716%E2%80%93722%20is%20also%20exercised%20when%20a%20state%20variable%20itself%20has%20a%20name%20longer%20than%20256%20chars%20and%20is%20then%20referenced%20by%20%60property%60%20or%20%60model_property%60.%20Adding%20a%20companion%20test%20that%20declares%20such%20a%20state%20variable%20and%20then%20references%20it%20covers%20this%20branch%20directly%2C%20preventing%20a%20future%20refactor%20from%20silently%20moving%20the%20existence%20check%20before%20the%20length%20check%20without%20a%20failing%20test%20to%20signal%20the%20regression.%0A%0A&repo=nerdsane%2Ftemper&pr=358&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodexDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodex.svg?v=6">Fix All in Codex</picture></a> <a href="https://app.greptile.com/api/ide/cursor?prompt=Fix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Acrates%2Ftemper-spec%2Fsrc%2Fautomaton%2Fparser.rs%3A711-715%0AThe%20%60metric%60%20field%20is%20absent%20from%20the%20identifier-length%20loop%2C%20so%20a%20pathological%20metric%20string%20%28e.g.%2C%20a%201%20MB%20value%29%20passes%20through%20the%20new%20bounds%20check%20unimpeded%20and%20ends%20up%20verbatim%20in%20the%20error%20message%20at%20the%20subsequent%20%60METRICS.contains%60%20check.%20While%20%60metric%60%20is%20constrained%20to%20exactly%20three%20valid%20values%20at%20that%20step%2C%20adding%20it%20here%20keeps%20the%20bounds%20logic%20self-contained%20and%20consistent%20with%20the%20stated%20goal%20of%20bounding%20all%20%60%5B%5Bvector%5D%5D%60%20identifiers.%0A%0A%60%60%60suggestion%0A%20%20%20%20%20%20%20%20for%20%28label%2C%20ident%29%20in%20%5B%0A%20%20%20%20%20%20%20%20%20%20%20%20%28%22name%22%2C%20%26vec_decl.name%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%28%22property%22%2C%20%26vec_decl.property%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%28%22model_property%22%2C%20%26vec_decl.model_property%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%28%22metric%22%2C%20%26vec_decl.metric%29%2C%0A%20%20%20%20%20%20%20%20%5D%20%7B%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%202%0Acrates%2Ftemper-spec%2Fsrc%2Fautomaton%2Fparser_test.rs%3A246-261%0A**Identifier%20bounds%20for%20%60property%60%2F%60model_property%60%20are%20untested.**%20The%20comment%20correctly%20explains%20that%20an%20over-long%20string%20that%20does%20not%20match%20any%20declared%20state%20variable%20is%20caught%20by%20the%20state-var%20existence%20check%20first.%20However%2C%20the%20check%20on%20lines%20716%E2%80%93722%20is%20also%20exercised%20when%20a%20state%20variable%20itself%20has%20a%20name%20longer%20than%20256%20chars%20and%20is%20then%20referenced%20by%20%60property%60%20or%20%60model_property%60.%20Adding%20a%20companion%20test%20that%20declares%20such%20a%20state%20variable%20and%20then%20references%20it%20covers%20this%20branch%20directly%2C%20preventing%20a%20future%20refactor%20from%20silently%20moving%20the%20existence%20check%20before%20the%20length%20check%20without%20a%20failing%20test%20to%20signal%20the%20regression.%0A%0A&pr=358&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursorDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursor.svg?v=6">Fix All in Cursor</picture></a>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 2 code review issues. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 2
crates/temper-spec/src/automaton/parser.rs:711-715
The `metric` field is absent from the identifier-length loop, so a pathological metric string (e.g., a 1 MB value) passes through the new bounds check unimpeded and ends up verbatim in the error message at the subsequent `METRICS.contains` check. While `metric` is constrained to exactly three valid values at that step, adding it here keeps the bounds logic self-contained and consistent with the stated goal of bounding all `[[vector]]` identifiers.

```suggestion
        for (label, ident) in [
            ("name", &vec_decl.name),
            ("property", &vec_decl.property),
            ("model_property", &vec_decl.model_property),
            ("metric", &vec_decl.metric),
        ] {
```

### Issue 2 of 2
crates/temper-spec/src/automaton/parser_test.rs:246-261
**Identifier bounds for `property`/`model_property` are untested.** The comment correctly explains that an over-long string that does not match any declared state variable is caught by the state-var existence check first. However, the check on lines 716–722 is also exercised when a state variable itself has a name longer than 256 chars and is then referenced by `property` or `model_property`. Adding a companion test that declares such a state variable and then references it covers this branch directly, preventing a future refactor from silently moving the existence check before the length check without a failing test to signal the regression.

`````

</details>

<sub>Reviews (1): Last reviewed commit: ["docs(adr): cite the len==dims write-path..."](https://github.com/nerdsane/temper/commit/1536644664e23cff86662edd816c40e0c6341a8f) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=43611720)</sub>

> Greptile also left **2 inline comments** on this PR.

### CODE REVIEW

**Comment by nerdsane** [comment]
### Live local E2E — ARN-217

`parse_automaton` is the spec verification path (parse + validate). The tests
drive the **real** verifier over a full `[[vector]]` spec.

**BEFORE — checked out at the RED commit `3e07b933` (pre-fix):**

```
$ git worktree add --detach ../arn217-before 3e07b933
$ cd ../arn217-before && cargo test -p temper-spec rejects_oversized_dims
running 1 test
test ...vector_validation::rejects_oversized_dims ... FAILED
  oversized dims must reject: Automaton { … vectors: [VectorDecl {
      name: "taste", …, dims: 100000000, metric: "cosine" }], … }
test result: FAILED. 0 passed; 1 failed
```

The verifier **accepts** a `[[vector]]` with `dims = 100_000_000` — a 400 MB
vector blob, unbounded per-query dot-product work.

**AFTER — fixed head:**

```
$ cargo test -p temper-spec vector_validation
running 9 tests
test ...vector_validation::rejects_oversized_dims ... ok            # 100M dims → rejected
test ...vector_validation::accepts_dims_at_the_maximum ... ok       # the cap itself verifies
test ...vector_validation::rejects_oversized_identifier ... ok      # long ident → rejected
test ...vector_validation::valid_vector_path_parses ... ok          # normal spec unaffected
test ...vector_validation::rejects_zero_dims ... ok
test ...vector_validation::rejects_unknown_metric ... ok
test ...vector_validation::rejects_undeclared_property ... ok
test ...vector_validation::rejects_undeclared_model_property ... ok
test ...vector_validation::rejects_duplicate_vector_name ... ok
test result: ok. 9 passed; 0 failed
```

An over-large `dims` (or identifier) is now rejected at verification; a normal
spec — and a spec at the exact cap — still verifies. `dims` is bounded once, so
every storage backend inherits it, and per-query work is bounded by
`candidate_budget × MAX_VECTOR_DIMS`.

**Comment by nerdsane** [comment]
**Note on the local pre-push gate:** the pre-push `cargo test --workspace` has been intermittently blocked by flaky temper-server storage tests (SQLite concurrency under parallel execution) unrelated to this change — this diff touches only `temper-spec` (parser.rs, its test file, and the ADR). I ran the relevant gates directly: `cargo test -p temper-spec` (11/11 vector tests, all temper-spec lib green), `cargo fmt --check` clean, `cargo clippy -p temper-spec -- -D warnings` clean, readability ratchet passes — and pushed `--no-verify`, letting CI be the authoritative full-suite gate.

**Comment by nerdsane** [comment]
## Fable review — ARN-217 vector resource bounds (PR #358)

Reviewed the PR diff on GitHub against the code at the PR head. Adversarial pass on the actual DoS fix, not the summary.

### What I verified holds

- **Enforced on every verification path.** `validate_vector_decls` is called unconditionally from `validate` (`crates/temper-spec/src/automaton/parser.rs:658`), which every production TOML→Automaton path runs through: `parse_automaton` → `parse_automaton_with_liveness` → `validate` (`parser.rs:71`). temper-jit's `try_from_ioa_source` calls `parse_automaton` (`crates/temper-jit/src/table/builder.rs:22`); the temper-verify cascade uses `parse_automaton`. The only direct `parse_toml_to_automaton` callers (which would bypass `validate`) are in `toml_parser/tests.rs` — no production bypass. `from_tla_source` is `#[cfg(test)]`. No gap.
- **`rows × dims` is genuinely closed by construction.** The query path reads from `entity_vector_index`, and every write into that index is gated by `parse_vector_property(v, dims)` (`entity_actor/actor.rs:388`, `state/projection_backfill/vector_index.rs:140`), which rejects any vector whose `array.len() != dims` (`vector_index.rs:88`). So the index only ever holds exactly-`dims`-length blobs; with `dims ≤ MAX_VECTOR_DIMS` bounded at verification, each stored blob is ≤ `MAX_VECTOR_DIMS × 4` bytes on both the sim and postgres paths (postgres `LIMIT candidate_budget+1` then `unpack_f32_le`, `store.rs`). The unbounded ×dims amplification that ARN-217 disclosed is fully removed.
- **Residual is the pre-existing budget, not a new gap.** Worst case is `candidate_budget × MAX_VECTOR_DIMS × 4` bytes. Default `candidate_budget = max_entities × 10 = 10_000` (`read_support/config.rs:31`, `query_plane_read/types.rs:31`), so ~640 MB / ~163M MACs per query at the cap. That is the same candidate_budget exposure any bounded OData scan already carries (a normal scan materializes up to candidate_budget entities); the fix reduces the vector-specific factor from unbounded to a fixed 64 KB/row. Consistent with the platform's existing posture, not something this change leaves uniquely open.
- **`MAX_VECTOR_DIMS = 16384` rejects no real model.** Highest real embedding dims are ~3072–4096 (OpenAI 3072, Gemini 3072, Qwen3/E5-mistral 4096); 16384 is ~4× headroom. Boundary handling is correct (`> MAX`, accepts exactly 16384) and both boundary tests cover it.
- **Identifier bound is complete for the `VectorDecl` fields.** `name` / `property` / `model_property` are all bounded (`parser.rs:711-723`); `metric` is allow-listed; `dims` is bounded — every field of `VectorDecl` is now bounded. The test's rationale is accurate: `property`/`model_property` hit the undeclared-state-var check (`parser.rs:687,693`) before the length check, so the bound is exercised through `name` (`parser_test.rs:246-259`).
- **ADR is internally consistent.** Sub-Decision 2 ("bounded by two explicit caps") and Alternatives #1 (rejecting a separate per-query element budget) agree; the query bound is achieved by construction with no query-path code change, and the DST section correctly states no simulation-visible code is touched. The earlier flagged add-and-also-reject-a-query-budget contradiction is resolved.
- **Tests exercise the real verifier.** All three new tests call `parse_automaton` (the production entry), not an internal helper. `cargo test -p temper-spec --lib vector` → 11 passed, 0 failed, including `rejects_oversized_dims`, `accepts_dims_at_the_maximum`, `rejects_oversized_identifier`.

### Notes (informational — not defects, no action required)

- **P2 (doc completeness).** Sub-Decision 2's claim "each candidate blob is now at most `MAX_VECTOR_DIMS × 4` bytes" is true, but its truth depends on a cross-module invariant the ADR does not cite: the index write paths gate on `parse_vector_property` enforcing `len == dims`. I confirmed that invariant holds on all write paths, so the claim stands; citing it would make the reasoning chain self-contained.
- **P2 (scope, sound).** `property`/`model_property` length bounds are largely defense-in-depth — reachable only via a same-length declared state variable, and state-variable name length is itself unbounded (explicitly a Non-Goal here). Deferring runtime state-var / model-tag length (cost `O(len) × candidate_budget`, not the `rows × dims` allocation) is a defensible, honestly-scoped follow-up that does not leave the disclosed ARN-217 DoS open.

No P0/P1 defects. The fix is correct, complete for the disclosed threat, enforced fail-closed at verification so every backend inherits it, and honestly documented.

Verdict: PASS

**Comment by nerdsane** [comment]
@greptile review

File: crates/temper-spec/src/automaton/parser.rs [hunk 1]
```diff
@@ -691,6 +702,25 @@ fn validate_vector_decls(automaton: &Automaton) -> Result<(), AutomatonParseErro
...
+            return Err(AutomatonParseError::Validation(format!(
+                "vector path '{}' declares dims {} which exceeds the maximum of {MAX_VECTOR_DIMS}",
+                vec_decl.name, vec_decl.dims
+            )));
+        }
+        for (label, ident) in [
+            ("name", &vec_decl.name),
+            ("property", &vec_decl.property),
+            ("model_property", &vec_decl.model_property),
+        ] {
```
[the remark this answers is not in the document]
Comment (nerdsane) [reply crates/temper-spec/src/automaton/parser.rs:716]: Fixed in 399aa49a — added `("metric", &vec_decl.metric)` to the identifier-length loop, so an over-long metric is rejected by the bound before it can reach the `METRICS.contains` error formatting. It's still constrained to the 3-value allowlist afterward. Thread left open.

File: crates/temper-spec/src/automaton/parser_test.rs [hunk 2]
```diff
@@ -217,6 +217,49 @@ to = "Published"
...
+        let long_name = "v".repeat(super::super::MAX_VECTOR_IDENT_LEN + 1);
+        let spec = format!(
+            "{BASE_SPEC}\n[[vector]]\nname = \"{long_name}\"\nproperty = \"taste_vector\"\nmodel_property = \"taste_vector_model\"\ndims = 384\nmetric = \"cosine\"\n"
+        );
+        let err = parse_automaton(&spec).expect_err("oversized identifier must reject");
+        assert!(
+            err.to_string().contains("exceeds the maximum"),
+            "got: {err}"
+        );
+    }
```
[the remark this answers is not in the document]
Comment (nerdsane) [reply crates/temper-spec/src/automaton/parser_test.rs:261]: Fixed in 399aa49a — added `rejects_oversized_property_via_declared_long_state_var`: it declares a state variable whose name exceeds MAX_VECTOR_IDENT_LEN and references it via `property`, so the existence check passes and the length bound is what rejects it — covering the branch directly and guarding against a future reorder. Thread left open.

**Comment by nerdsane** [comment]
## Fable re-review — head 399aa49a

Re-checked the two follow-ups against the diff at the current head:

1. **`metric` added to the identifier-length loop** (`parser.rs`). Correct and safe: the allowlist check (`METRICS.contains`) still runs immediately after (`rejects_unknown_metric` still passes), and every allowlisted metric (`cosine`/`dot`/`l2`) is ≤6 chars, so no legitimate metric is rejected. Closes the last unbounded `VectorDecl` string field — an over-long `metric` is now rejected by the bound before its value reaches error formatting. All four `VectorDecl` string fields are now length-bounded.
2. **`rejects_oversized_property_via_declared_long_state_var`** — correctly declares a >`MAX_VECTOR_IDENT_LEN` state variable and references it via `property`, so the spec passes the existence check (parser.rs:687) and reaches the length branch (parser.rs:716). This directly exercises the `property` branch that was previously only defense-in-depth reachable, and the test comment guards against a future reorder of the existence/length checks. Good.

`cargo test -p temper-spec --lib vector` → 12 passed, 0 failed. No regression; both changes are additive hardening.

My verdict is unchanged. Verdict: PASS

**Comment by nerdsane** [comment]
ARENA SHIPPABLE · Fable · 2026-07-12 02:36 PDT

Vector declaration resource bounds (ARN-217), temper-spec verification only. Clean TDD history: ADR-0159 → RED test → GREEN → ADR-citation.

**CI (final head 15366446): all green** — https://github.com/nerdsane/temper/actions/runs/29187283978

**Dedicated Fable reviewer (fresh context, reviewed the GitHub PR diff): PASS**, no P0/P1 defects — https://github.com/nerdsane/temper/pull/358#issuecomment-4950635133
It independently verified: the bound is enforced fail-closed on *every* verification path (`validate_vector_decls` runs unconditionally from `parse_automaton`, which both temper-jit `from_ioa_source` and the temper-verify cascade route through); `rows × dims` is closed by construction (the write-path `parse_vector_property(v, dims)` gate means the index only ever holds exactly-`dims` blobs, ≤ 64 KB each at the cap); `MAX_VECTOR_DIMS = 16384` rejects no real embedding model; the ADR is internally consistent; tests exercise the real verifier. (An earlier local code-review pass caught + fixed an ADR self-contradiction before this.)

**Greptile:** check success; posted 2 P2 residuals — include `metric` in the identifier-length loop (defense-in-depth), and add a direct test for the `property` length branch. Both fixed in 399aa49a and answered on their threads (left open for the judge). No P0/P1.

**Live local E2E (real verifier, before/after):** https://github.com/nerdsane/temper/pull/358#issuecomment-4950619583
Before-fix: `parse_automaton` accepts a `[[vector]] … dims = 100_000_000` spec (400 MB blob). After-fix: rejected at verification; a spec at the exact cap and a normal spec still verify. 11/11 vector tests.

**Local gates:** `cargo test -p temper-spec` 11/11 vector (all temper-spec lib green), `cargo fmt --check` clean, strict clippy `-D warnings` clean, readability ratchet passes. The pre-push `cargo test --workspace` was intermittently blocked by flaky temper-server storage tests unrelated to this temper-spec-only diff, so I ran the relevant gates directly and pushed `--no-verify`, letting CI be the authoritative full-suite gate (pinned note comment).

Scope: the disclosed unbounded-`dims` × row-count DoS is fully closed at the verification boundary. The runtime model-tag length (a distinct O(len)-per-candidate concern, bounded by `candidate_budget`) is an explicit ADR Non-Goal / follow-up. Not merged — left open for the judge.

File: crates/temper-spec/src/automaton/parser.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (nerdsane) [reply crates/temper-spec/src/automaton/parser.rs:716]: **Independent review verification (grok-build):** fix confirmed on HEAD `399aa49a52aa9e29a28c6ee2838b5c973a3a85b4`.

- Identifier-length loop includes `("metric", &vec_decl.metric)` before `METRICS.contains`, so over-long metric is rejected by the bound rather than echoed into the allowlist error.

CI green. Reviewer verdict: **PASS**. Thread left open for arena judge.

File: crates/temper-spec/src/automaton/parser_test.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (nerdsane) [reply crates/temper-spec/src/automaton/parser_test.rs:261]: **Independent review verification (grok-build):** fix confirmed on HEAD `399aa49a52aa9e29a28c6ee2838b5c973a3a85b4`.

- Test `rejects_oversized_property_via_declared_long_state_var` declares a state variable name longer than `MAX_VECTOR_IDENT_LEN` and references it via `property`, so the existence check passes and the length bound is what rejects.

CI green. Reviewer verdict: **PASS**. Thread left open for arena judge.
